perf-tool/src/infra.cpp

```diff
@@ -49,6 +49,55 @@ bool check_jvmti_error(jvmtiEnv *jvmti, jvmtiError errnum, const char *str) {
     return true;
 }
 
+void EventConfig::setCallbacks(const std::string& cls, const std::string& method, const std::string& sig)
+{
+    std::lock_guard<std::mutex> lg(configMutex);
+    callbackClass.assign(cls);
+    callbackMethod.assign(method);
+    callbackSignature.assign(sig);
+    /* Also reset the cached IDs; they need to be re-discovered */
+    if (callbackIDs.cachedCallbackClass)
+        jni_env->DeleteGlobalRef(callbackIDs.cachedCallbackClass);
+    callbackIDs.cachedCallbackClass = NULL;
+    callbackIDs.cachedCallbackMethodId = NULL;
+}
+
+EventConfig::CallbackIDs EventConfig::getCallBackIDs(JNIEnv *env)
+{
+    std::lock_guard<std::mutex> lg(configMutex);
+    if (!getCachedCallbackClass() || !getCachedCallbackMethodId())
+    {
+        /* Need to compute them */
+        if (!getCallbackClass().empty() && !getCallbackMethod().empty() && !getCallbackSignature().empty())
+        {
+            jclass callbackCls = env->FindClass(getCallbackClass().c_str());
```

Comment:
I think you need to add the following:
```
if (env->ExceptionOccurred(env))
{
   env->ExceptionDescribe(env);
   env->ExceptionClear(env);
}
```
because `FindClass` can throw an exception. Technically there's other places we call `FindClass` and maybe those need this check as well, but we we're mostly using it on JCL classes, whereas this is going to be a user defined class so it's possible the system loader may not find it.

Reply:
Good find. I just re-used the pattern from the other places without looking up the description of the API.
Interestingly, my small test program ran ok when I provided a class that did not exit, but we should definitely check for and suppress exceptions being propagated to Java level.

Reply:
Fixed